support-figma/extended-layout-plugin/src/image-res-module.ts

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { json } from "stream/consumers";
+import * as Utils from "./utils";
+
+const IMAGE_HASH_TO_RES_KEY = "image_hash_to_res";
+
+// Hash to image bytes map
+const imageBytesMap = new Map<string, Uint8Array>();
+// Hash to nodes array map
+const imageHashToNodesMap = new Map<string, Array<string>>();
+// Hash to res name map
+const imageHashToResMap = new Map<string, string>();
+// Res name to image hash map, this will be used to look up a unique res name.
+const resToImageHashMap = new Map<string, string>();
+
+export async function exportAllImagesAsync() {
+  await figma.loadAllPagesAsync();
+
+  imageBytesMap.clear();
+  imageHashToNodesMap.clear();
+  imageHashToResMap.clear();
+  resToImageHashMap.clear();
+
+  let cachedHashToResMap = load();
+
+  for (let page of figma.root.children) {
+    for (let child of page.children) {
+      await exportNodeImagesAsync(child, cachedHashToResMap);
+    }
+  }
+
+  save();
+
+  figma.showUI(__html__, { width: 600, height: 600 });
+  figma.ui.postMessage({
+    msg: "image-export",
+    imageBytesArray: Array.from(imageBytesMap),
+    imageNodesArray: Array.from(imageHashToNodesMap),
+    imageResNameArray: Array.from(imageHashToResMap),
+  });
+}
+
+export function updateResName(imageHash: string, resName: string) {
+  if (!resToImageHashMap.has(resName)) {
+    const oldName = imageHashToResMap.get(imageHash);
+    if (oldName) {
+      resToImageHashMap.delete(oldName);
+    }
+    imageHashToResMap.set(imageHash, resName);
+    resToImageHashMap.set(resName, imageHash);
+    save();
+  } else {
+    figma.notify("Res name already exists. Reverting....");
+  }
+
+  figma.ui.postMessage({
+    msg: "image-export",
+    imageBytesArray: Array.from(imageBytesMap),
+    imageNodesArray: Array.from(imageHashToNodesMap),
+    imageResNameArray: Array.from(imageHashToResMap),
+  });
+}
+
+async function exportNodeImagesAsync(
+  node: SceneNode,
+  cachedHashToResMap: Map<string, string>
+) {
+  let nodeWithFills: GeometryMixin = node as GeometryMixin;
```

Comment:
What is a GeometryMixin?

Reply:
`interface GeometryMixin extends MinimalStrokesMixin, MinimalFillsMixin`. It is like the base interface for nodes both have strokes and fills.
I broke it into MinimalStrokesMixin and MinimalFillsMixin.